cpp/src/arrow/io/caching.cc

```diff
@@ -171,7 +171,8 @@ Status ReadRangeCache::Cache(std::vector<ReadRange> ranges) {
   }
 
   impl_->AddEntries(std::move(entries));
-  return Status::OK();
+  // Prefetch immediately, regardless of executor availability, if possible
+  return impl_->file->WillNeed(ranges);
```

Comment:
`ReadAsync` by default only queues the read for execution in a thread pool. It doesn't necessarily issue the read immediately. So this might still be useful, though perhaps only in uninteresting corner cases.

Reply:
Hey, sorry for the delay. I think what Francois proposes with the file readers calling only `WillNeed` makes sense. That way consumers can mostly remain the same. However for (parallel) caching to work under this API, this means some file implementations will become stateful, which may be confusing/error-prone, so I think there's some more consideration warranted for the Parquet path.